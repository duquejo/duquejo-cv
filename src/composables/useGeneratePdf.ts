import { computed } from 'vue';
import type { Ref } from 'vue';
import type { GenerateProps } from '@pdfme/common';
import type { MessageSchema } from '@/i18n/config';
import { generateInputs, generateBaseTemplate } from '@/components/pdf/config';

const loadPdfGenerator = () => import('@pdfme/generator');

const useGeneratePdf = (locale: Ref<string>, languageSources: Ref<MessageSchema>) => {

  const template = computed(() => generateBaseTemplate(locale));
  const inputs = computed(() => generateInputs(languageSources));

  const createPdf = async () => {

    const config: GenerateProps = {
      inputs: inputs.value,
      template: await template.value,
    };

    const { generate } = await loadPdfGenerator();

    generate(config).then((pdf) => {
      // Browser
      const blob = new Blob([pdf.buffer], { type: 'application/pdf' });
      window.open(URL.createObjectURL(blob));
    });
  }

  return {
    createPdf,
  };
}

export default useGeneratePdf;