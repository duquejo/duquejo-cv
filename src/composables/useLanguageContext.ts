import { ref, watch } from 'vue';
import { useI18n } from 'vue-i18n';

import CONFIG_CO from '../shared/locale/es-CO/config.json';
import CONFIG_US from '../shared/locale/en-US/config.json';

export enum SupportedLanguages {
  ES_CO = 'es-CO',
  EN_US = 'en-US',
};

type SourceType = Record<SupportedLanguages, any>;

const SOURCES: SourceType = {
  [SupportedLanguages.ES_CO]: CONFIG_CO,
  [SupportedLanguages.EN_US]: CONFIG_US,
}

const useLanguageContext = () => {
  const { t, locale } = useI18n({ useScope: 'global' });
  const languageSources = ref(SOURCES[locale.value as SupportedLanguages]);
  watch(locale, (val: string) => {
    switch (val) {
      case SupportedLanguages.ES_CO:
        return languageSources.value = SOURCES[SupportedLanguages.ES_CO];
      case SupportedLanguages.EN_US:
        return languageSources.value = SOURCES[SupportedLanguages.EN_US];
      default:
        console.error('Language not supported.')
        break;
    }
  });

  return {
    t,
    languageSources,
  };
};

export default useLanguageContext;