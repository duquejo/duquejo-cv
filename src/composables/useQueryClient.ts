import { useQuery } from '@tanstack/vue-query';
import githubApi from '@/api/githubApi';
import { Event } from '@/interfaces/github-events';

const getEvents = async () => {
  try {
    const { data } = await githubApi.get<Event[]>('/events?per_page=10&page=1');
    return data;
  } catch (error) {
    console.error(error);
  }
};

const useQueryClient = () => {

  const { isLoading, data: events } = useQuery({
    queryKey: ['events'],
    queryFn: getEvents,
    staleTime: 1000 * 60 * 60,
  });

  return {
    events,
    isLoading,
  };
};

export default useQueryClient;