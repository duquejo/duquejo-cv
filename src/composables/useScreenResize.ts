import { DESKTOP_BREAKPOINT } from '@/shared/constants';

const useScreenResize = () => {
  return {
    isMobile: () => screen.width <= DESKTOP_BREAKPOINT
  };
};

export default useScreenResize;