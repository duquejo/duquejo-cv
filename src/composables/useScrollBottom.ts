import { ref } from 'vue';
import { PIXELS_GAP } from '@/shared/constants';
const useScrollBottom = () => {

  const isBottomReached = ref<boolean>(false);

  const handleScroll = (target: HTMLElement) => {
    const { scrollTop, scrollHeight, clientHeight } = target;
    const calc = Math.round(scrollTop + clientHeight) >= Math.round(scrollHeight - PIXELS_GAP);
    isBottomReached.value = calc;
  }
  return {
    isBottomReached,
    handleScroll,
  };
};

export default useScrollBottom;