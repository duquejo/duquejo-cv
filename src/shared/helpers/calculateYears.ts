import { computed } from 'vue';
import { START_ADVENTURE_YEAR } from '@/shared/constants';

export const calculateYears = computed(() => (new Date().getFullYear() - START_ADVENTURE_YEAR));