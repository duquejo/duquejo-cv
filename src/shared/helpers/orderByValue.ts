
export const orderObjectListByKey = <T>( list: T[], key: keyof T ): T[] | null => {
  if( ! list || list.length === 0 ) return null;
  return list.sort((a: T, b: T) => parseInt(String(b[key])) - parseInt(String(a[key])) );
}