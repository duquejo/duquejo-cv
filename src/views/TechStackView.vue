<template>
  <WrapperLayout>
    <section>
      <h1 class="text-h1 leading-tight">
        {{ t('general.services.title') }} & <br />{{ t('general.services.title2') }}
      </h1>
    </section>
    <section>
      <!-- General Services -->
      <h2 class="text-h2 hidden md:block">{{ t('general.services.subtitle') }}</h2>
      <div class="grid grid-cols-1 md:grid-cols-2 gap-x-8 gap-y-6">
        <IconCardComponent v-for="({ content, title, icon }) in languageSources.services" :content="content"
          :title="title" :icon="icon" />
      </div>
    </section>
    <section class="row">
      <div class="basis-full">

        <!-- Experience -->
        <h3 class="text-h3 mb-6 mt-5">{{ t('general.services.experience.title') }}</h3>
        <TimelineComponent
          v-for="({ role, type, additional_info, date_range, enterprise, isRecent, project, resume }, index) in languageSources.experience"
          :role="role" :type="(type as Type)" :additional_info="additional_info" :date_range="date_range"
          :enterprise="enterprise" :is-recent="isRecent" :project="project" :resume="resume" :key="index" />

        <!-- Education -->
        <h3 class="text-h3 mb-6 mt-10">{{ t('general.services.education.title') }}</h3>
        <TimelineComponent
          v-for="({ role, type, additional_info, date_range, enterprise, isRecent, project, resume }, index) in languageSources.education"
          :role="role" :type="(type as Type)" :additional_info="additional_info" :date_range="date_range"
          :enterprise="enterprise" :is-recent="isRecent" :project="project" :resume="resume" :key="index" />

      </div>
      <div class="basis-1/2">

        <!-- Idiomas -->
        <h3 class="text-h3 mb-6 mt-5">{{ t('general.services.languages.title') }}</h3>
        <div class="grid grid-cols-2 gap-x-8 gap-y-6">
          <IconCardComponent v-for="({ content, title, icon }) in languageSources.languages" :content="content"
            :title="title" :icon="icon" :order="Order.Col" />
        </div>

        <!-- Languages -->
        <h3 class="text-h3 mb-6 mt-10">{{ t('general.services.tech_languages.title') }}</h3>
        <ProgressBarComponent v-for="(language, i) in orderedLanguages" :key="`${language.tag}-${i}`"
          :color="language.color" :tag="language.tag" :value="language.value" />

        <!-- Framework & Libraries -->
        <h3 class="text-h3 mb-6 mt-10">{{ t('general.services.framelibs.title') }}</h3>
        <ProgressBarComponent v-for="(framelib, i) in orderedFrameLibs" :key="`${framelib.tag}-${i}`"
          :color="framelib.color" :tag="framelib.tag" :value="framelib.value" />

        <!-- Knowledges -->
        <h3 class="text-h3 mb-6 mt-10">{{ t('general.services.knowledges.title') }}</h3>
        <div class="flex flex-wrap gap-2 lg:gap-3 justify-evenly lg:justify-normal">
          <PillComponent v-for="(knowledge, i) in languageSources.knowledges" :key="`${knowledge}-${i}`"
            :text="knowledge" />
        </div>

        <!-- Certifications -->
        <h3 class="text-h3 mb-6 mt-10">{{ t('general.services.certifications.title') }}</h3>
        <div class="grid grid-cols-1 md:grid-cols-3  lg:grid-cols-1 gap-4">
          <CCardComponent />
        </div>
      </div>
    </section>
  </WrapperLayout>
</template>

<script setup lang="ts">
import { computed } from 'vue';

import { Order, Type, Bars } from '../interfaces';

import useLanguageContext from '@/composables/useLanguageContext';

import WrapperLayout from '@/layouts/WrapperLayout.vue';
import TimelineComponent from '@/components/timeline/TimelineComponent.vue';
import CCardComponent from '@/components/certification-card/CCardComponent.vue';
import ProgressBarComponent from '@/components/progress-bar/ProgressBarComponent.vue';
import PillComponent from '@/components/pill/PillComponent.vue';
import IconCardComponent from '@/components/card/IconCardComponent.vue';

import { orderObjectListByKey } from '@/shared/helpers/orderByValue';

const { t, languageSources } = useLanguageContext();

const orderedLanguages = computed(() => orderObjectListByKey<Bars>(languageSources.value.main_languages, 'value'));
const orderedFrameLibs = computed(() => orderObjectListByKey<Bars>(languageSources.value.framelibs, 'value'));
</script>