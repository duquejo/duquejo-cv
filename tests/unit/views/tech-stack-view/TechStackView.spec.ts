import { mount } from "@vue/test-utils";
import { describe, expect, it, vi } from "vitest";

import TechStackView from '../../../../src/views/TechStackView.vue';
import createVueI18n from "../../../mocks/mock-i18n";

import * as orderByValue from '../../../../src/shared/helpers/orderByValue';
import * as useLanguageContext from '../../../../src/composables/useLanguageContext';

describe('+ TechStackView tests', () => {

  const i18n = createVueI18n('en');

  const orderByValueSpy = vi.spyOn(orderByValue, 'orderObjectListByKey');
  const useLanguageContextSpy = vi.spyOn(useLanguageContext, 'default');

  const scrollIntoViewSpy = vi.fn();
  window.HTMLElement.prototype.scrollIntoView = scrollIntoViewSpy;

  it('Should match the snapshot', async () => {
    const wrapper = mount(TechStackView, {
      global: {
        plugins: [i18n],
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(orderByValueSpy).toHaveBeenCalled();
    expect(orderByValueSpy).toHaveBeenCalledTimes(2);
    
    expect(useLanguageContextSpy).toHaveBeenCalled();

    expect(scrollIntoViewSpy).toHaveBeenCalled();
    expect(scrollIntoViewSpy).toHaveBeenCalledWith({
      behavior: 'instant'
    });
  });
});