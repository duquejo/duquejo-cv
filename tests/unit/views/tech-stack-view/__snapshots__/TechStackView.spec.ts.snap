// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ TechStackView tests > Should match the snapshot 1`] = `
"<div class="animated-container">
  <article>
    <section>
      <h1 class="text-h1 leading-tight">general.services.title &amp; <br>general.services.title2</h1>
    </section>
    <section>
      <!-- General Services -->
      <h2 class="text-h2 hidden md:block">general.services.subtitle</h2>
      <div class="grid grid-cols-1 md:grid-cols-2 gap-x-8 gap-y-6"></div>
    </section>
    <section class="row">
      <div class="basis-full">
        <!-- Experience -->
        <h3 class="text-h3 mb-6 mt-5">general.services.experience.title</h3><!-- Education -->
        <h3 class="text-h3 mb-6 mt-10">general.services.education.title</h3>
      </div>
      <div class="basis-1/2">
        <!-- Idiomas -->
        <h3 class="text-h3 mb-6 mt-5">general.services.languages.title</h3>
        <div class="grid grid-cols-2 gap-x-8 gap-y-6"></div><!-- Languages -->
        <h3 class="text-h3 mb-6 mt-10">general.services.tech_languages.title</h3><!-- Framework & Libraries -->
        <h3 class="text-h3 mb-6 mt-10">general.services.framelibs.title</h3><!-- Knowledges -->
        <h3 class="text-h3 mb-6 mt-10">general.services.knowledges.title</h3>
        <div class="flex flex-wrap gap-2 lg:gap-3 justify-evenly lg:justify-normal"></div><!-- Certifications -->
        <h3 class="text-h3 mb-6 mt-10">general.services.certifications.title</h3>
        <div class="grid grid-cols-1 md:grid-cols-3 lg:grid-cols-1 gap-4"><a href="#" class="group flex flex-col items-center border border-gray-200 rounded-lg shadow xl:flex-row xl:max-w-xl hover:bg-gray-100 hover:text-gray-600 overflow-hidden gap-4 transition-colors"><img class="object-cover w-full rounded-t-lg max-h-40 xl:rounded-none xl:rounded-s-lg grayscale group-hover:grayscale-0 transition-all xl:max-h-40" src="/src/assets/avatar-small.jpg" alt="Certification">
            <div class="flex flex-col justify-between leading-normal mr-4">
              <h5 class="text-sm font-semibold tracking-tight xl:mt-2 mx-4 xl:mx-auto">Noteworthy technology acquisitions 2021</h5>
              <p class="m-4 xl:mx-auto">Lorem ipsum, dolor sit amet.</p>
            </div>
          </a></div>
      </div>
    </section>
  </article>
  <footer>
    <ul class="flex gap-x-4">
      <li class="icon"><a href="https://github.com/duquejo01" target="_blank" title="Github">
          <v-icon name="bi-github" scale="1"></v-icon>
        </a></li>
      <li class="icon"><a href="https://www.linkedin.com/in/duquejo" target="_blank" title="LinkedLn">
          <v-icon name="co-linkedin-in" scale="1"></v-icon>
        </a></li>
    </ul><span>©️ 2024 José Duque</span>
  </footer>
</div>"
`;
